metricbeat/module/windows/services/services_windows.go

```diff
@@ -0,0 +1,374 @@
+		for {
+			if err := _EnumServicesStatusEx(handle, ScEnumProcessInfo, ServiceWin32, state, &servicesBuffer[0], bufSize, &bytesNeeded, &servicesReturned, nil, nil); err != nil {
+				if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_MORE_DATA {
+					return nil, ServiceErrno(err.(syscall.Errno))
+				}
+				bufSize += bytesNeeded
+			} else {
+
+				for i := 0; i < int(servicesReturned); i++ {
+					serviceTemp := (*EnumServiceStatusProcess)(unsafe.Pointer(&servicesBuffer[i*sizeofEnumServiceStatusProcess]))
+
+					service, err := getServiceInformation(serviceTemp, servicesBuffer, lastOffset, handle)
+					if err != nil {
+						return nil, err
+					}
+
+					services = append(services, service)
+				}
+
+				return services, nil
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func getServiceInformation(rawService *EnumServiceStatusProcess, servicesBuffer []byte, lastOffset uintptr, handle ServiceDatabaseHandle) (ServiceStatus, error) {
+	service := ServiceStatus{}
+
+	displayNameBuffer := new(bytes.Buffer)
+	serviceNameBuffer := new(bytes.Buffer)
+
+	serviceNameOffset := uintptr(unsafe.Pointer(rawService.LpServiceName)) - (uintptr)(unsafe.Pointer(&servicesBuffer[0]))
+	displayNameOffset := uintptr(unsafe.Pointer(rawService.LpDisplayName)) - (uintptr)(unsafe.Pointer(&servicesBuffer[0]))
+
+	displayNameBuffer.Reset()
+	serviceNameBuffer.Reset()
+
+	if err := sys.UTF16ToUTF8Bytes(servicesBuffer[displayNameOffset:serviceNameOffset], displayNameBuffer); err != nil {
+		return service, err
+	}
+
+	if err := sys.UTF16ToUTF8Bytes(servicesBuffer[serviceNameOffset:lastOffset], serviceNameBuffer); err != nil {
+		return service, err
+	}
+
+	lastOffset = displayNameOffset
+
+	service.DisplayName = displayNameBuffer.String()
+	service.ServiceName = serviceNameBuffer.String()
+
+	var state string
+
+	if stat, ok := serviceStates[ServiceState(rawService.ServiceStatusProcess.DwCurrentState)]; ok {
+		state = stat
+	} else {
+		state = "Can not define State"
+	}
+	service.CurrentState = state
+
+	// Get detailed information
+	if err := getDetailedServiceInfo(handle, service.ServiceName, ServiceQueryConfig, &service); err != nil {
+		return service, err
+	}
+
+	//Get uptime for service
+	if ServiceState(rawService.ServiceStatusProcess.DwCurrentState) != ServiceStopped {
+		processUpTime, err := getServiceUptime(rawService.ServiceStatusProcess.DwProcessId)
+		if err != nil {
+			logp.Warn("Uptime for service %v is not available", service.ServiceName)
+		}
+		service.Uptime = processUpTime
+	}
+
+	return service, nil
+}
+
+// getServiceUptime returns the uptime for process
+func getServiceUptime(processID uint32) (time.Duration, error) {
+	var processCreationTime gosigar.ProcTime
+
+	err := processCreationTime.Get(int(processID))
+	if err != nil {
+		return time.Duration(processCreationTime.StartTime), err
+	}
+
+	tm := time.Now().UnixNano()
+	uptime := time.Duration((tm / int64(time.Millisecond)) - int64(processCreationTime.StartTime))
```

Comment:
I'm not sure if this is any more clear, but there is the `time.Since()` function for the purpose of calculating elapsed times.

`uptime := time.Since(time.Unix(0, int64(processCreationTime.StartTime) * int64(time.Millisecond)))`

Reply:
Thanks!